core/src/apps/cardano/address.py

```diff
@@ -34,6 +39,39 @@
 MAX_ADDRESS_BYTES_LENGTH = 65
 
 
+def validate_address_parameters(parameters: CardanoAddressParametersType) -> None:
+    if parameters.address_type == CardanoAddressType.BYRON:
+        if not is_byron_path(parameters.address_n):
+            raise INVALID_ADDRESS_PARAMETERS
+    elif parameters.address_type in ADDRESS_TYPES_SHELLEY:
+        if not is_shelley_path(parameters.address_n):
+            raise INVALID_ADDRESS_PARAMETERS
+
+        if parameters.address_type == CardanoAddressType.BASE:
+            _validate_base_address_staking_info(
+                parameters.address_n_staking, parameters.staking_key_hash
+            )
+        elif parameters.address_type == CardanoAddressType.POINTER:
+            if parameters.certificate_pointer is None:
+                raise INVALID_ADDRESS_PARAMETERS
+        elif parameters.address_type == CardanoAddressType.REWARD:
+            if not SCHEMA_STAKING_ANY_ACCOUNT.match(parameters.address_n):
+                raise INVALID_ADDRESS_PARAMETERS
+    else:
+        raise INVALID_ADDRESS_PARAMETERS
+
+
+def _validate_base_address_staking_info(
+    staking_path: list[int],
+    staking_key_hash: bytes | None,
+) -> None:
+    if (staking_key_hash is None) == (not staking_path):
```

Comment:
I'd rather break this down into two separate condition asserting that neither both nor none can be present

Reply:
Perhaps we don't need to do a validation of both of them being present? If you pass in `staking_key_hash` that is the one that's going to be used. Otherwise we validate and then also use `staking_path`?

I remember this specific condition being recommended by someone from Trezor in their review of the Shelley update, but I think we can separate and see if they accept it - if we decide to keep it.

Reply:
If we don't validate the case of them both being present, then the whole function would be more readable. And I also noticed, that we should probably check the length of `staking_key_hash`.

Reply:
I agree it should be rewritten to be more readable. `staking_key_hash` naturally splits it into two cases to handle.

Reply:
I've updated the `if`s. It now doesn't check whether both `staking_key_hash` and `staking_path` are set. If `staking_key_hash` is set, that is the one that's used.

I've also added a check for `staking_key_hash` length and a test for that.